yelp_scrape.py

```diff
@@ -23,16 +24,30 @@ def url_generator(search_item, location):
     search_item = search_item.replace(' ', '+')
     location = location.replace(' ', '+').replace(',', '%2C')
     url = base_url + 'find_desc=' + search_item + '&find_loc=' + location
-    return url
-
-
-search, loc = get_search_request()
-yelp_page = requests.get(url_generator(search, loc))
-yelp_soup = BeautifulSoup(yelp_page.content, 'html.parser')
-
-test = yelp_soup.find_all('a')
-links = []
-for a in test:
-    links.append(a["href"])
-
-print(links)
+    file_name = search_item + location
+    return url, file_name
+
+# local_cache_check will take the search request url and check a local directory for a file of the same name
+# if the file is found then it will use that file instead of scraping the internet
+def local_cache_check(url, file_name):
+    # url_hash = hash(url)
+    path = input('Please enter complete filepath to search: ')
+    file_list = os.listdir(path)
+    file_name = file_name + '.html'
+    file_path = os.path.join(path, file_name)
+    if file_name in file_list:
+        print('Retrieving the file from the directory')
+        with open(file_path, 'r') as file:
+            yelp_soup = file.read()
```

Comment:
A problem you're probably running into is that `file.read()` returns a `str`
whereas under the `else` block, `yelp_soup` is a `BeautifulSoup` object

I see why this might be unintuitive, you're writing the BeautifulSoup object to the file
the problem is that files don't know anything about Python, you can basically write either strings or raw bytes
so `BeautifulSoup.prettify()` returns a `str` and then you write that string to file

so when you read it back and return it, it's a string, not a BeautifulSoup object

this is why type hints are extremely useful

if you annotate this function as `def local_cache_check(url: str, file_name: str | Path) -> BeautifulSoup:`

then your VSCode window will yell at you on this line, saying "hey this function is supposed to return a `BeautifulSoup`, but you're returning a `str`, why u do dis?

without type annotations, it's really easy to return the wrong type like you're doing here

Reply:
Can you explain the `str | Path` piece of that?

Reply:
`str | Path` means it's _either_ a `str` or a `Path`, the function will accept either one
it's useful for when you don't actually care about certain aspects of a type
for example, you can often treat a `list` and a `set` the same way, and so a lot of functions could take a `list | set`, or more generally, an `Iterable` I think